source/common/filter/http/filter_config_discovery_impl.cc

```diff
@@ -196,24 +197,62 @@ std::shared_ptr<FilterConfigSubscription> FilterConfigProviderManagerImpl::getSu
 }
 
 FilterConfigProviderPtr FilterConfigProviderManagerImpl::createDynamicFilterConfigProvider(
-    const envoy::config::core::v3::ConfigSource& config_source,
-    const std::string& filter_config_name, const std::set<std::string>& require_type_urls,
-    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
-    bool apply_without_warming) {
-  auto subscription =
-      getSubscription(config_source, filter_config_name, factory_context, stat_prefix);
+    const envoy::config::core::v3::ExtensionConfigSource& config_source,
+    const std::string& filter_config_name, Server::Configuration::FactoryContext& factory_context,
+    const std::string& stat_prefix) {
+  auto subscription = getSubscription(config_source.config_source(), filter_config_name,
+                                      factory_context, stat_prefix);
   // For warming, wait until the subscription receives the first response to indicate readiness.
   // Otherwise, mark ready immediately and start the subscription on initialization. A default
   // config is expected in the latter case.
-  if (!apply_without_warming) {
+  if (!config_source.apply_default_config_without_warming()) {
     factory_context.initManager().add(subscription->initTarget());
   }
-  auto provider = std::make_unique<DynamicFilterConfigProviderImpl>(
-      std::move(subscription), require_type_urls, factory_context);
+  std::set<std::string> require_type_urls;
+  for (const auto& type_url : config_source.type_urls()) {
+    auto factory_type_url = TypeUtil::typeUrlToDescriptorFullName(type_url);
+    require_type_urls.emplace(factory_type_url);
+  }
+  auto provider = std::make_unique<DynamicFilterConfigProviderImpl>(subscription, require_type_urls,
+                                                                    factory_context);
   // Ensure the subscription starts if it has not already.
-  if (apply_without_warming) {
+  if (config_source.apply_default_config_without_warming()) {
     factory_context.initManager().add(provider->init_target_);
   }
+
+  // If the subscription already received a config, attempt to apply it.
+  bool config_applied = false;
+  if (subscription->lastConfig().has_value()) {
+    try {
+      provider->validateConfig(subscription->lastTypeUrl());
+      provider->onConfigUpdate(subscription->lastConfig().value(), subscription->lastVersionInfo(),
+                               nullptr);
+      config_applied = true;
+    } catch (const EnvoyException& e) {
+      ENVOY_LOG(debug, "ECDS subscription {} is invalid in a listener context: {}.",
```

Comment:
Thinking through this more broadly, I kind of find this corner case a bit icky. This relates to my validate config comment above; if all we can do is validate configs, can we pull in the type URL validation into the core here?

Reply:
Are you making the argument here that since the last config has already been validated and is known to work, the only possible way it could fail here is if the type URLs mismatch somehow? Do we even need `onConfigUpdate` in try/catch if we validate?

In terms of stats, probably something in https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/stats#config-http-conn-man-stats namespace makes sense.

Reply:
@htuch I'm looking for guidance how to deal with inconsistency between ECDS and LDS. The scenario is the following:
1. LDS add listener L1 with config_discovery for x with type URL {T1, T2}
2. ECDS update for x with type URL T1
3. LDS add listener L2 with config_discovery for x with type URL T2.

I think this is similar to a delayed NACK situation since we want the management server to push a new ECDS update for x with type URL T2, that validates both L1 and L2. This PR will make L2 to use the default config, but it does not indicate to the management server than a new ECDS update is needed. Granted, this is a fairly convoluted example. The metric would be useful, but is the assumption here, that some higher level controller (human?) is going to look for it and force push ECDS?

Reply:
Added "config_conflict" stat for ECDS resource. I think keying by ECDS is a bit more actionable, since it is clear why resource is the conflict (LDS tends to be more source-of-truth). LMK.

Reply:
In your example, is "x" actually different in step 4? I.e. it has been deliberately changed by the management server and when server in hypothetical step 4, it's really just a problem of 3 and 4 racing? Or is "x" somehow polymorphic and could be T1 or T2 and it gets decided based on client state?

This seems a pretty weird situation but I'm likely missing something. Feel free to ping me IM so we can figure this out with lower RTT.